internal/zero/analytics/storage.go

```diff
@@ -0,0 +1,116 @@
+package analytics
+
+import (
+	"context"
+	"encoding/base64"
+	"fmt"
+	"time"
+
+	"google.golang.org/protobuf/types/known/anypb"
+	"google.golang.org/protobuf/types/known/structpb"
+
+	"github.com/pomerium/pomerium/pkg/grpc/databroker"
+	"github.com/pomerium/pomerium/pkg/protoutil"
+)
+
+const (
+	metricStateTypeURL = "pomerium.io/ActiveUsersMetricState"
+)
+
+// SaveMetricState saves the state of a metric to the databroker
+func SaveMetricState(
+	ctx context.Context,
+	client databroker.DataBrokerServiceClient,
+	id string,
+	data []byte,
+	lastReset time.Time,
+) error {
+	_, err := client.Put(ctx, &databroker.PutRequest{
+		Records: []*databroker.Record{{
+			Type: metricStateTypeURL,
+			Id:   id,
+			Data: (&MetricState{
+				Data:      data,
+				LastReset: lastReset,
+			}).ToAny(),
+		}},
+	})
+	return err
+}
+
+// LoadMetricState loads the state of a metric from the databroker
+func LoadMetricState(
+	ctx context.Context, client databroker.DataBrokerServiceClient, id string,
+) (*MetricState, error) {
+	resp, err := client.Get(ctx, &databroker.GetRequest{
+		Type: metricStateTypeURL,
+		Id:   id,
+	})
+	if err != nil {
+		return nil, fmt.Errorf("load metric state: %w", err)
+	}
+
+	var state MetricState
+	err = state.FromAny(resp.GetRecord().GetData())
+	if err != nil {
+		return nil, fmt.Errorf("load metric state: %w", err)
+	}
+
+	return &state, nil
+}
+
+// MetricState is the persistent state of a metric
+type MetricState struct {
+	Data      []byte
+	LastReset time.Time
+}
+
+const (
+	dataKey      = "data"
+	lastResetKey = "last_reset"
+)
+
+// ToAny marshals a MetricState into an anypb.Any
+func (r *MetricState) ToAny() *anypb.Any {
+	return protoutil.NewAny(&structpb.Struct{
+		Fields: map[string]*structpb.Value{
+			dataKey:      structpb.NewStringValue(base64.StdEncoding.EncodeToString(r.Data)),
```

Comment:
Do protobuf strings not support arbitrary data?

More generally, is there a reason to prefer an ad-hoc Struct proto to a defined message type? Do we anticipate needing to evolve this data type in the future?

Reply:
> Do protobuf strings not support arbitrary data?
doesn't seem like it https://protobuf.dev/reference/protobuf/google.protobuf/#value

> is there a reason to prefer an ad-hoc Struct proto to a defined message type? 

I can try that, but then it'll be not in `internal/zero`

Reply:
In general, for the databroker, I kinda wish all the existing types were structs. The reason why is that to unmarshal an `Any` type, you need that type's protobuf definition to be imported. This makes it much harder for consumers of Databroker data to work with arbitrary data.